### ISSUE TITLE
META - Backward Compatibility

### ISSUE THREAD

**Issue by piux2** [gnolang/gno#1896]
We can use this as a starting point to track potential backward compatibility issues we need to be aware of, and devise solutions for each along the way.

### Context
It's important to be aware that all native functions could have backward compatibility issues when we update their native implementation. This is not introduced by the new binding solution; instead, it makes the potential issue more apparent.

During contract execution, the functionValue.nativeBody, https://github.com/gnolang/gno/blob/831bb6f92e1a2217242169dab1f4fd1f87e5eaa0/gnovm/pkg/gnolang/op_call.go#L63
which is a func(*Machine) returned from iterating through nativeFuncs in stdlibs/native.go, by finding matching package and  names, 
https://github.com/gnolang/gno/blob/831bb6f92e1a2217242169dab1f4fd1f87e5eaa0/gnovm/stdlibs/native.go#L26
returns a matching func(*Machine) when the VM executes the contract. 

For example, in the standard contract stdlib/std/native.gno, PrevRealm() > getRealm(1), the nativeFuncs links Gno's getRealm(1) 
https://github.com/gnolang/gno/blob/831bb6f92e1a2217242169dab1f4fd1f87e5eaa0/gnovm/stdlibs/std/native.gno#L57
with the native implementation of X_getRealm().
https://github.com/gnolang/gno/blob/831bb6f92e1a2217242169dab1f4fd1f87e5eaa0/gnovm/stdlibs/std/native.go#L70
https://github.com/gnolang/gno/blob/831bb6f92e1a2217242169dab1f4fd1f87e5eaa0/gnovm/stdlibs/native.go#L535

### Backward compatibility issues:  

updating the logic in stdlibs/std/native.go could cause existing contracts to fail to execute if contract developers assume the behaviors of PrevRealm have changed in a new release.

**Comment by zivkovicmilos** [comment]
cc @ajnavarro

**Comment by ajnavarro** [comment]
## Storage related backward-incompatibility problems

### Serialization System Changes

Our backward compatibility is tied to the serialization format used to store data on disk. Any change to this format could introduce backward incompatibilities.

### Payload Size Changes in `Set` Operations

Optimizations in the size of payloads stored during Set operations could lead to backward incompatibilities. Suppose the size of the data being stored in subsequent versions of the VM (even if that data is representing the same internal VM state, but is downsized using compression or similar) is reduced or altered for efficiency.

### State format change

Any change to the internal VM state format could break backward compatibility. For instance, introducing a garbage collector (GC) or altering the way the VM represents its internal state could modify the amount of data stored.

## Proposal

To mitigate the risk of backward-incompatible changes, I propose we decouple the storage costs from the exact byte size stored by the VM. Instead, each type—such as Interfaces, Structs, Ints, Strings, etc.—should be assigned a fixed cost based on properties intrinsic to the object itself.

For example, in the case of strings, we could calculate storage costs based on the length of the string (i.e., string length * X). This would ensure that the cost calculation is stable and not influenced by changes to how the VM stores data internally.

**Comment by Kouteki** [comment]
Conclusion of the team discussion:

1. Use best judgement on a case-by-case basis
2. Breaking compatility is not a deal breaker
3. We'll rely on chain upgrades and snapshots when it comes to it
4. YOLO

**Comment by thehowl** [comment]
> It's important to be aware that all native functions could have backward compatibility issues when we update their native implementation. This is not introduced by the new binding solution; instead, it makes the potential issue more apparent.

Yes, furthermore they need to be reimplemented by any and all alternative VM implementations.

Native bindings are a necessary evil which make our lives simpler for writing static analysis tools, as we don't need to add shims or hardcode Gno's standard library functions, but they should still be kept to be as little as possible.

**Comment by thehowl** [comment]
One general comment, additionally;

let's keep track of potential backward compatibility pitfalls, yes, but I'm looking more to have:

- Strong tests which strictly verify our current expected behaviour.
- #2159 can help us try typical user flows and ensure they always work.
- #2869 tries to prevent failures restarting the portal loop; I think eventually on mainnet we should have a long test (similar, conceptually, to this issue) which tries to replay the entire chain up to the point of the test with the modified code, showing us if the chain would fail with the updated code.

A collection like this can tell us where to look; proper tooling can help us point to real problems before they happen.

**Comment by Kouteki** [comment]
We addressed the storage topic in #1838 so I created https://github.com/gnolang/gno/issues/3519 for the native function.

I'll close the meta issue.
